neuralpp/test/quick_tests/symoblic/z3_usage_test.py

```diff
@@ -97,3 +98,67 @@ def test_sum():
     assert (is_valid(Sum([j for j in range(N+1)]) == (1 + N)*N/2))
     assert (is_valid(Sum([x for _ in range(N+1)]) == (1 + N)*x))
     assert (is_valid(Sum([x + j for j in range(N+1)]) == (1 + N)*x + (1 + N)*N/2))
+
+
+def test_z3_solver():
+    """
+    We want to test if Z3's solver environment can be copied. The use case:
+    say we already have constraint C, and we see a new literal L; we want to check the satisfiability of both
+    "C and L" and "C and not L"; if we can clone the Z3 environment/solver, then we can treat it as immutable.
+    Otherwise, we must treat it as mutable and can only check "C and L", call pop(), and then check "C and not L".
+    """
+    # the use case of using only one Solver()
+    s = Solver()
+    x, y = Ints('x y')
+    constraints = And(x > 2, y < 0)
+    s.add(constraints)
+    literal = x > y
+    s.push()  # create a new scope, so z3 knows where it should pop() to
```

Comment:
I am not catching the meaning of "knows where it should pop() to", can you please clarify?